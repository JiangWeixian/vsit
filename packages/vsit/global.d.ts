import type { Encode, Hook } from 'console-feed'
import type { ViteDevServer } from 'vite'

declare global {
  // eslint-disable-next-line vars-on-top, no-var
  var __rpc: {
    send: (channel: string, log: any) => void
  }
  // eslint-disable-next-line vars-on-top, no-var
  var __viteDevServer: ViteDevServer | undefined
  // eslint-disable-next-line vars-on-top, no-var
  var __hook: Hook
  // eslint-disable-next-line vars-on-top, no-var
  var __encode: Encode
}

export {}
