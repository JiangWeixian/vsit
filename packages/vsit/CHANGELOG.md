# vsit

## 0.0.4

### Patch Changes

- 7db8bfe: set max width of editor to 50%
- 6cd8d0a: support persist cache in node side
- 420a517: service worker cache in browser side
- b28891d: fix edit web content not working

## 0.0.3

### Patch Changes

- 99f03bd: replace @nicksrandall/console-feed to console-feed

## 0.0.2

### Patch Changes

- 29ddadc: support ts
- 29ddadc: use @nicksrandall/console-feed related to https://github.com/samdenty/console-feed/issues/124
- d20459e: common debug log

## 0.0.1

### Patch Changes

- 412c21f: mvp version
