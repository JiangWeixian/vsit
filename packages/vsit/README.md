# @aiou/bin-template
*build npm cli command application*

[![npm](https://img.shields.io/npm/v/@aiou/bin-template)](https://github.com/spring-catponents/bin-template) [![GitHub](https://img.shields.io/npm/l/@aiou/bin-template)](https://github.com/spring-catponents/bin-template) [![stackblitz](https://img.shields.io/badge/%E2%9A%A1%EF%B8%8Fstackblitz-online-blue)](https://stackblitz.com/github/spring-catponents/bin-template)

[Edit on StackBlitz ⚡️](https://stackblitz.com/github/spring-catponents/bin-template)

## features

- Native ESM Module development
- Use [**cac**](https://github.com/cacjs/cac) build cli application
- Interaction interface with [**inquirer**](https://github.com/SBoudrias/Inquirer.js/) and [**ora**](https://github.com/sindresorhus/ora)
- Type safe
- Release with github workflows and changeset
- pnpm
- vitest

## install 

```console
pnpm i -g @aiou/bin-template
```

## commands

### `hello`

say hello world with select option

`bin-template hello [word]`

### `loading`

display loading and loading text

`bin-template loading --text=[text] [ms]`

## development

- `pnpm run build`
