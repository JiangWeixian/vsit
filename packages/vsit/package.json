{
  "name": "vsit",
  "version": "0.2.2",
  "packageManager": "pnpm@7.33.0",
  "description": "todo",
  "keywords": ["vsit"],
  "license": "MIT",
  "engines": {
    "node": ">=16"
  },
  "homepage": "https://github.com/vsit#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vsit.git",
    "directory": "packages/vsit"
  },
  "bugs": {
    "url": "https://github.com/vsit/issues",
    "email": "jiangweixian1994@gmail.com"
  },
  "author": {
    "name": "JW",
    "email": "jiangweixian1994@gmail.com",
    "url": "https://twitter.com/jiangweixian"
  },
  "files": [
    "dts",
    "dist",
    "dist-client",
    "vendors"
  ],
  "bin": {
    "vsit": "bin/index.mjs"
  },
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "node": {
        "import": "./dist/node.mjs",
        "require": "./dist/node.cjs",
        "types": "./dts/exports/node.d.ts"
      },
      "browser": {
        "import": "./dist/browser.mjs",
        "require": "./dist/browser.cjs",
        "types": "./dts/exports/browser.d.ts"
      }
    }
  },
  "main": "dist/node.cjs",
  "module": "dist/node.mjs",
  "types": "dts/exports/node.d.ts",
  "scripts": {
    "build:dts": "rimraf dts && tspc -p tsconfig.dts.json",
    "build:lib": "rimraf dist && rollup -c --silent --environment BUILD:production",
    "build": "run-s build:*",
    "dev": "rollup -c -w",
    "preview": "node ./bin/index.mjs",
    "vitest": "vitest",
    "test": "pnpm run build && pnpm run vitest",
    "prepublishOnly": "publish-police",
    "ci:publish": "pnpm run build && pnpm changeset publish",
    "ci:version": "pnpm changeset version",
    "ci:snapshot": "pnpm changeset version --snapshot beta",
    "ci:prerelease": "pnpm run build && pnpm changeset publish --tag beta"
  },
  "dependencies": {
    "rs-module-lexer": "^2.3.0",
    "sirv": "^2.0.3",
    "vite": "^4.3.8"
  },
  "devDependencies": {
    "@changesets/cli": "^2.26.0",
    "@rollup/plugin-alias": "4.0.3",
    "@rollup/plugin-commonjs": "24.0.1",
    "@rollup/plugin-json": "^6.0.0",
    "@rollup/plugin-node-resolve": "15.0.1",
    "@rollup/plugin-replace": "^5.0.2",
    "@types/debug": "^4.1.8",
    "@types/fs-extra": "^11.0.1",
    "@types/inquirer": "^8.1.3",
    "@types/lodash-es": "^4.17.8",
    "@types/node": "20.3.1",
    "@types/rimraf": "^3.0.2",
    "body-parser": "1.20.2",
    "cac": "6.7.12",
    "consola": "^2.15.3",
    "console-feed": "3.5.0",
    "cz-emoji": "^1.3.1",
    "debug": "^4.3.4",
    "esbuild": "^0.18.6",
    "execa": "^6.0.0",
    "fs-extra": "^11.1.1",
    "husky": "^8.0.3",
    "inquirer": "8.2.0",
    "lodash-es": "^4.17.21",
    "npm-run-all": "^4.1.5",
    "ofetch": "^1.0.1",
    "ora": "6.0.1",
    "picocolors": "1.0.0",
    "publish-police": "^0.1.0",
    "read-yaml-file": "2.1.0",
    "rimraf": "^3.0.2",
    "rollup": "3.19.1",
    "rollup-plugin-condition-exports": "2.0.0-next.3",
    "rollup-plugin-esbuild": "^5.0.0",
    "rollup-plugin-node-externals": "5.1.2",
    "rollup-plugin-size": "^0.3.1",
    "source-map-support": "^0.5.21",
    "tempy": "^3.1.0",
    "ts-patch": "^3.1.2",
    "type-fest": "^3.12.0",
    "typescript": "5.1.6",
    "typescript-transform-paths": "^3.4.6",
    "ufo": "^1.1.2",
    "vitest": "^0.22.1",
    "vsit-shared": "workspace:*",
    "write-yaml-file": "5.0.0"
  }
}
