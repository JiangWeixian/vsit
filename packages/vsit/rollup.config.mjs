import path from 'node:path'

import alias from '@rollup/plugin-alias'
import commonjs from '@rollup/plugin-commonjs'
import json from '@rollup/plugin-json'
import resolve from '@rollup/plugin-node-resolve'
import replace from '@rollup/plugin-replace'
import { defineConfig } from 'rollup'
import ce from 'rollup-plugin-condition-exports'
import esbuild from 'rollup-plugin-esbuild'
import { externals } from 'rollup-plugin-node-externals'
import size from 'rollup-plugin-size'
import { createRequire } from 'node:module'

const require = createRequire(import.meta.url)

const plugins = [
  externals({
    devDeps: false,
    builtinsPrefix: 'add',
  }),
  commonjs(),
  replace({
    delimiters: ['', ''],
    preventAssignment: true,
    values: {
      'import \'source-map-support/register.js\';': '',
    },
  }),
  esbuild({
    minify: false,
    sourceMap: process.env.BUILD !== 'production',
    target: 'es2021',
  }),
  alias({
    customResolver: resolve({ extensions: ['.tsx', '.ts'] }),
    entries: Object.entries({
      '@/*': ['./src/*'],
    }).map(([alias, value]) => ({
      find: new RegExp(`${alias.replace('/*', '')}`),
      replacement: path.resolve(process.cwd(), `${value[0].replace('/*', '')}`),
    })),
  }),
  resolve({ browser: false, exportConditions: ['node', 'default'] }),
  json(),
  size(),
]

export default defineConfig([
  {
    input: {
      'write-yaml-file/index': require.resolve('write-yaml-file')
    },
    plugins,
    output: [
      {
        entryFileNames: '[name].cjs',
        dir: 'vendors',
        chunkFileNames: 'chunks/[name].cjs',
        format: 'cjs',
      },
    ],
  },
  {
    input: 'src/cli.ts',
    preserveEntrySignatures: 'strict',
    external: ['source-map-support/register.js'],
    plugins,
    watch: {
      exclude: ['./package.json'],
    },
    output: [
      {
        sourcemap: process.env.BUILD !== 'production',
        entryFileNames: '[name].mjs',
        dir: 'dist',
        chunkFileNames: 'chunks/[name].mjs',
        format: 'esm',
      },
    ],
  },
  {
    plugins: [
      ce({
        outDir: 'dist',
        declarationDir: 'dts',
      }),
      ...plugins,
    ],
    watch: {
      exclude: ['./package.json'],
    },
    output: [
      {
        sourcemap: true,
        entryFileNames: '[name].mjs',
        dir: 'dist',
        chunkFileNames: 'chunks/[name].mjs',
        format: 'esm',
      },
      {
        sourcemap: true,
        entryFileNames: '[name].cjs',
        dir: 'dist',
        chunkFileNames: 'chunks/[name].cjs',
        format: 'cjs',
      },
    ],
  },
])
