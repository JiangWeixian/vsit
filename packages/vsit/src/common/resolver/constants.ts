export const ESMSH_HTTP_RE = /(^https?:\/\/?esm\.sh)/
export const ESMSH_HTTP_SUB_RE = /(^\/v[0-9]+\/)/
export const ESMSH_PROTOCOL_RE = /^esm\.sh:/

export const ESM_HOST = 'https://esm.sh/'
export const ESMSH_PROTOCOL = 'esm.sh:'

export const VIRTUAL_RE = /^(?:virtual:)|\0/

export const VIRUTAL_NODE_ID = 'fake-node-file'
export const RESOLVED_NODE_ID = `\0${VIRUTAL_NODE_ID}`
