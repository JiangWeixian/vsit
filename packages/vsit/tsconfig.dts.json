{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "plugins": [
      { "transform": "typescript-transform-paths" },
      {
        "transform": "typescript-transform-paths",
        "afterDeclarations": true
      }
    ],
    "rootDir": "src"
  },
  "exclude": ["node_modules", "lib", "test"],
  "include": ["./global.d.ts", "src"]
}
