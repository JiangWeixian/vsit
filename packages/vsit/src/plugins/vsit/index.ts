import { performance } from 'node:perf_hooks'

import bodyparser from 'body-parser'
import Hook from 'console-feed/lib/Hook'
import { Encode } from 'console-feed/lib/Transform'
import { parseURL } from 'ufo'
import {
  API_PARSE_IMPORTS,
  NODE_API_PATH,
  WBE_API_PATH,
} from 'vsit-shared/constants'

import { debug } from '@/common/log'
import { parseImports } from '@/common/parse'
import { VIRUTAL_NODE_ID, VIRUTAL_WEB_ID } from '@/common/resolver/constants'
import { isEsmSh } from '@/common/resolver/is'
import { createStore } from '@/common/store'
import {
  parseModulesDeps,
  transform,
  unwrapId,
  wrapCode,
  wrapId,
} from '@/common/utils'

import type { AsyncReturnType } from 'type-fest'
import type { Plugin, ViteDevServer } from 'vite'

const invalid = async (moduleName: string, server: ViteDevServer) => {
  const module = await server.moduleGraph.getModuleByUrl(moduleName)
  module && server.moduleGraph.invalidateModule(module)
}

export const PluginVit = (): Plugin[] => {
  let nodeContent = ''
  let webContent = ''
  let store: AsyncReturnType<typeof createStore>
  return [
    {
      name: 'vsit:core',
      async configResolved() {
        store = await createStore()
      },
      configureServer(server) {
        // TODO: common middlewares or standalone vite plugin
        globalThis.__viteDevServer = server
        globalThis.__encode = Encode
        // globalThis.__decode = Decode
        globalThis.__hook = Hook
        server.middlewares.use(bodyparser.json())
        server.middlewares.use(async (req, res, next) => {
          const url = parseURL(req.url)
          if (url.pathname === API_PARSE_IMPORTS && req.method === 'POST') {
            const body = req.body as unknown as { code: string; filename: string }
            const imports = parseImports(body.code, body.filename)
            res.end(JSON.stringify(imports))
            return
          }
          if (url.pathname === WBE_API_PATH && req.method === 'POST') {
            webContent = req.body.content
            res.end('ok')
            return
          }
          if (url.pathname === WBE_API_PATH && req.method === 'GET') {
            // Invalid before transformRequest, make sure transformRequest get latest content from user
            await invalid(VIRUTAL_WEB_ID, server)
            webContent = (await server.transformRequest(VIRUTAL_WEB_ID, { ssr: false }))?.code ?? ''
            res.setHeader('Content-Type', 'text/javascript')
            res.end(webContent)
            return
          }
          if (url.pathname === NODE_API_PATH && req.method === 'POST') {
            const body = req.body
            nodeContent = body.content
            debug.plugin('update fake node file %s', nodeContent)
            res.end(nodeContent)
            return
          }
          if (url.pathname === NODE_API_PATH && req.method === 'GET') {
            try {
              // /fake-node-file?t=<timestamp>
              await server.ssrLoadModule(VIRUTAL_NODE_ID)
              const module = await server.moduleGraph.getModuleByUrl(VIRUTAL_NODE_ID)
              const packages = parseModulesDeps(module)
              store.cache.writePackages(packages)
              // console.log([...module?.ssrTransformResult?.values()][1], [...module?.importedModules?.values()][1].ssrTransformResult?.deps)
              if (module) {
                res.setHeader('Content-Type', 'text/javascript')
                res.end(module.ssrTransformResult?.code ?? '')
              } else {
                res.end('ok')
              }
              invalid(VIRUTAL_NODE_ID, server)
            } catch (e) {
              console.error(e)
            }
            return
          }
          next()
        })
      },
      resolveId(id, importer) {
        console.log('importer', importer)
        // Should not add \0 to id, tell vite to transform it to js(if VIRUTAL_NODE_ID is ts)
        if (id === VIRUTAL_NODE_ID) {
          return VIRUTAL_NODE_ID
        }
        if (id === VIRUTAL_WEB_ID) {
          return VIRUTAL_WEB_ID
        }
        return null
      },
      async load(id) {
        if (id === VIRUTAL_NODE_ID) {
          return transform(nodeContent, 'node')
        }
        if (id === VIRUTAL_WEB_ID) {
          return transform(webContent, 'web')
        }
        return null
      },
    },
    {
      name: 'vsit:esmsh',
      resolveId(id) {
        if (isEsmSh(id)) {
          const resolvedId = wrapId(id)
          debug.plugin('resolved esmsh id from %s to %s', id, resolvedId, resolvedId[0])
          return resolvedId
        }
        return null
      },
      async load(id, options) {
        // vite will remove duplicate slash if id starts with 'https://'
        if (isEsmSh(id)) {
          if (!options?.ssr) {
            debug.plugin('skip load esmsh id %s on non-ssr mode', id)
            return ''
          }
          // un wrap
          const now = performance.now()
          const url = unwrapId(id)
          const code = await store.fetch(url)
          const resolvedCode = wrapCode(code)
          debug.benchmark('load url %s took', url, `${(performance.now() - now) / 1000}ms`)
          return {
            code: resolvedCode,
            moduleSideEffects: false,
          }
        }
        return null
      },
    },
  ]
}
