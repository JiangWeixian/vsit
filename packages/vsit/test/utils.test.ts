import {
  describe,
  expect,
  it,
} from 'vitest'

import { isEsmSh } from '@/common/resolver/is'
import {
  injectConsoleHook,
  parseDeps,
  transform,
  unwrapId,
  wrapCode,
  wrapId,
} from '@/common/utils'

describe('string', () => {
  it('wrap id', () => {
    expect(wrapId('esm.sh:react')).toBe('\0https://esm.sh/react')
    expect(wrapId('/v126/react')).toBe('\0https://esm.sh/v126/react')
  })

  it('unwrap id(maybe virtual) to url', () => {
    expect(unwrapId('esm.sh:react')).toBe('https://esm.sh/react')
    expect(unwrapId('https:/esm.sh/react')).toBe('https://esm.sh/react')
    expect(unwrapId('https://esm.sh/react')).toBe('https://esm.sh/react')
    expect(unwrapId('\0esm.sh:react')).toBe('https://esm.sh/react')
    expect(unwrapId('\0https:/esm.sh/react')).toBe('https://esm.sh/react')
    expect(unwrapId('/@id/__x00__https:/esm.sh/lodash-es@4.17.21')).toBe('https://esm.sh/lodash-es@4.17.21')
  })

  it('wrap code', () => {
    expect(wrapCode(`
import { uniq } from "https://esm.sh/lodash-es@4.17.21"
import stripAnsi from "https://esm.sh/strip-ansi@7.1.0"
    `)).toMatchSnapshot()
  })

  it('inject console hook on node file', () => {
    expect(injectConsoleHook('console.log("injected")')).toMatchInlineSnapshot(`
    "
    import { consolehook } from \\"vsit\\"
    globalThis.__hook(consolehook, (log) => {
      console.log(log)
      globalThis.__rpc.send(log)
    })
    console.log(\\"injected\\")
    "
    `)
  })

  describe('transform', () => {
    it('node file', () => {
      expect(transform('console.log("injected")', 'node')).toMatchInlineSnapshot(`
      "
      import { consolehook } from \\"vsit\\"
      globalThis.__hook(consolehook, (log) => {
        console.log(log)
        globalThis.__rpc.send(log)
      })
      consolehook.log(\\"injected\\")
      "
      `)
    })
    it('web file', () => {
      expect(transform('console.log("injected")', 'web')).toMatchInlineSnapshot('"consolehook.log(\\"injected\\")"')
    })
  })
})

describe('parse', () => {
  it('parse deps', () => {
    const deps = parseDeps([
      '/@fs/path/node.mjs',
      '/@id/__x00__https:/esm.sh/lodash-es@4.17.21',
      '/@id/__x00__https:/esm.sh/strip-ansi@7.1.0',
    ])
    expect(deps).toMatchObject([
      'https://esm.sh/lodash-es@4.17.21',
      'https://esm.sh/strip-ansi@7.1.0',
    ])
  })
})

describe('utils', () => {
  it('is esm.sh', () => {
    expect(isEsmSh('\0https://esm.sh/react')).toBe(true)
    expect(isEsmSh('https://esm.sh/react')).toBe(true)
    expect(isEsmSh('https:/esm.sh/react')).toBe(true)
    expect(isEsmSh('esm.sh:react')).toBe(true)
    expect(isEsmSh('/v126/react')).toBe(true)
    expect(isEsmSh('/v126/ansi-regex@6.0.1/esnext/ansi-regex.mjs')).toBe(true)
  })
})
