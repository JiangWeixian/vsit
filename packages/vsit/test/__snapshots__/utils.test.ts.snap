// Vitest Snapshot v1

exports[`string > wrap code 1`] = `
"
import { uniq } from \\"esm.sh:lodash-es@4.17.21\\"
import stripAnsi from \\"esm.sh:strip-ansi@7.1.0\\"
    "
`;
